pkg/container/queue/chunkqueue.go

```diff
@@ -0,0 +1,389 @@
+	return q.chunks[q.head]
+}
+
+func (q *ChunkQueue[T]) lastChunk() *chunk[T] {
+	return q.chunks[q.tail-1]
+}
+
+// NewChunkQueue creates a new ChunkQueue
+func NewChunkQueue[T any]() *ChunkQueue[T] {
+	return NewChunkQueueLeastCapacity[T](1)
+}
+
+// NewChunkQueueLeastCapacity creates a ChunkQueue with an argument minCapacity.
+// It requests that the queue capacity be at least minCapacity. And it's similar
+// to the cap argument when making a slice using make([]T, len, cap)
+func NewChunkQueueLeastCapacity[T any](minCapacity int) *ChunkQueue[T] {
+	elementSize := unsafe.Sizeof(*new(T))
+	if elementSize == 0 {
+		log.Error("Cannot create a queue of type")
+		return nil
+	}
+
+	chunkLength := int(defaultSizePerChunk / elementSize)
+	if chunkLength < minimumChunkLen {
+		chunkLength = minimumChunkLen
+	}
+
+	q := &ChunkQueue[T]{
+		head:        0,
+		tail:        0,
+		size:        0,
+		chunkLength: chunkLength,
+	}
+	q.chunkPool = sync.Pool{
+		New: func() any {
+			return newChunk[T](q.chunkLength, q)
+		},
+	}
+
+	q.chunks = make([]*chunk[T], defaultPitchArrayLen, defaultPitchArrayLen)
+	q.extend(minCapacity)
+	return q
+}
+
+// Size returns the number of elements in queue
+func (q *ChunkQueue[T]) Size() int {
+	return q.size
+}
+
+// Len returns the number of elements in queue
+func (q *ChunkQueue[T]) Len() int {
+	return q.size
+}
+
+// Cap returns the capacity of the queue. The queue can hold more elements
+// than that number by automatic expansion
+func (q *ChunkQueue[T]) Cap() int {
+	return q.chunkLength*(q.tail-q.head) - q.chunks[q.head].l
+}
+
+// Empty indicates whether the queue is empty
+func (q *ChunkQueue[T]) Empty() bool {
+	return q.size == 0
+}
+
+// At returns the value of a given index. At() does NOT support modifying the value
+func (q *ChunkQueue[T]) At(idx int) (T, bool) {
+	if idx < 0 || idx >= q.size {
+		return q.defaultValue, false
+	}
+	i := q.chunks[q.head].l + idx
+	return q.chunks[q.head+i/q.chunkLength].data[i%q.chunkLength], true
+}
+
+// Replace assigns a new value to a given index
+func (q *ChunkQueue[T]) Replace(idx int, val T) bool {
+	if idx < 0 || idx >= q.size {
+		return false
+	}
+	i := q.chunks[q.head].l + idx
+	q.chunks[q.head+i/q.chunkLength].data[i%q.chunkLength] = val
+	return true
+}
+
+// Head returns the value of the first element. This method is only for reading
+// the first element, not for modification
+func (q *ChunkQueue[T]) Head() (T, bool) {
+	if q.Empty() {
+		return q.defaultValue, false
+	}
+	c := q.firstChunk()
+	return c.data[c.l], true
+}
+
+// Tail returns the value of the last element. This method is only for reading
+// the last element, not for modification
+func (q *ChunkQueue[T]) Tail() (T, bool) {
+	if q.Empty() {
+		return q.defaultValue, false
+	}
+	c := q.lastChunk()
+	return c.data[c.r-1], true
+}
+
+// extend extends the space by adding chunk(s) to the queue
+func (q *ChunkQueue[T]) extend(n int) {
+	if n <= 0 {
+		n = 1
```

Comment:
I don't know. Does this make sense? Because it's like an invalid input. Maybe we shouldn't modify it implicitly. But I am not an API expert, so I don't know what we should do.

Reply:
Maybe a no-op is more suitable for invalid input